### PR TITLE
Add ManageIQ alerts module

### PR DESCRIPTION
**Pull request by elad661**
##### SUMMARY
ManageIQ is an open source management platform for Hybrid IT.
This change is adding:

manageiq_alerts module, responsible for alerts management in ManageIQ

All the modules, including docs, tests and usage examples can be found here

requirements: manageiq-api-client-python
For "hash expression" support, https://github.com/ManageIQ/manageiq-api/pull/76 is required on the ManageIQ server. Versions of ManageIQ without this change will still work with this module, but only for alerts with a "miq_expression" and not a "hash_expression"

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
manageiq_alerts.py (module)

##### ANSIBLE VERSION
```
ansible 2.5.0
```

### CODE REVIEW

**Comment by elad661** [comment]
cc @yaacov @cben - please review.

**Review by yaacov — APPROVED** [review]
Great, shipit

p.s.
@elad661 , I will re-shipit again after you commit the re-factor that will remove the need for the `@staticmethod` decorator.
Don't worry about this getting merged before the re-factor is finished, we will do a new PR if needed.

File: lib/ansible/modules/remote_management/manageiq/manageiq_alert_policies.py
Comment (yaacov) [inline lib/ansible/modules/remote_management/manageiq/manageiq_alert_policies.py]: I like better a list comprehensions here:

```
keys = ['description', 'db', 'enabled', 'options', 'hash_expression']
existing_policy_dict = {key: existing_policy[key] for key in keys if key in existing_policy}
```

File: lib/ansible/modules/remote_management/manageiq/manageiq_alert_policies.py
Comment (yaacov) [inline lib/ansible/modules/remote_management/manageiq/manageiq_alert_policies.py]: description can be in multiple lines, I think it's nicer (same in line 46):

```
- absent - alert policy should not exist,
- present - alert policy should exist,
- list - return a list of alert policies.
```

File: lib/ansible/modules/remote_management/manageiq/manageiq_alert_policies.py
Comment (yaacov) [inline lib/ansible/modules/remote_management/manageiq/manageiq_alert_policies.py]: choices can be in multiple lines, I think it's nicer:
```

choices: ['Vm', 'ContainerNode', 'MiqServer', 'Host', 'Storage', 'EmsCluster',
    'ExtManagementSystem', 'MiddlewareServer']
```

File: lib/ansible/modules/remote_management/manageiq/manageiq_alert_policies.py
Comment (yaacov) [inline lib/ansible/modules/remote_management/manageiq/manageiq_alert_policies.py]: the long msg lines are not nice, try to find a way to shorten them.

for example:
```
msg_template = "Alert policy {description} created {status}: {details}"
msg_data = dict(description=policy['description'], status='successfully', details=result)

msg = msg_template.format(msg_data)
```

File: lib/ansible/modules/remote_management/manageiq/manageiq_alert_policies.py
Comment (yaacov) [inline lib/ansible/modules/remote_management/manageiq/manageiq_alert_policies.py]: nice ! I forgot about this one :-)

**Review by yaacov — APPROVED** [review]
LGTM 👍

**Comment by yaacov** [comment]
shipit 🐈

**Comment by cben** [comment]
I still haven't read these in-depth, but some naming/scope questions:

1. the term "alert policy" is not used anywhere in ManageIQ.  UI has:
  - Policy Profiles 
  - Policies
  - Alert Profiles
  - Alerts

  The REST api has `alert_definitions` for Alerts, in order to reserve `alerts` to mean "instance of Alert fired at specific moment" (MiqAlertStatus internally).
  It's a mess, but I'm not sure adding "alert policy" will improve it  :-), especially since "Policy" already means something in UI and docs, and *an Alert is not a Policy*

We now have 3 modules on the table:
- manageiq_policies (plural) #31233
  currently assigns existing(?) Policy Profiles to existing entities (e.g. provider), but intention was to later cover assigning Policies to Policy Profiles, and possibly defining Policy Profiles and/or Policies themselves — hence the catch-all name...
- manageiq_alert_policies this PR #32136 
  defines Alerts.
- manageiq_alert_profiles #32354
  defines Alert Profiles, and assigns existing Alerts to Alert Profiles.

Ideally, we'd have similar interfaces to policies vs alerts...

(don't rush to rename, that's a lot of potentially redudant work, just wanted to start the discussion, while I procrastinate on actually reviewing ;-))

cc @moolitayer @joelddiaz for naming advice

**Comment by elad661** [comment]
You're right about the name, I was sure I read/heard the term "alert policies" somewhere last week, but it's not in the UI and I can't find a reference to it anywhere right now.
So this module should probably be renamed to manageiq_alerts. Policies are something completely different.

**Comment by joelddiaz** [comment]
I think manageiq_alerts is a fine name. It fits with the UI.

I did try to run this module against my MIQ installation, an it fails when trying to 'absent' an already absent alert.

TASK [delete non-existing alert] ********************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Alert policy 'This alert should not exist' does not exist in ManageIQ"}

Since Ansible modules are supposed to be idempotent/rerunnable, a call to a non-existent alert should just return as 'change: false' since the alert is already absent.

**Comment by elad661** [comment]
@joelddiaz Okay, I fixed this and now it won't fail when the alert doesn't exist

**Comment by joelddiaz** [comment]
@elad661 i can't get the example 'add an alert' example to work. I get the following error message:

```
TASK [Add an alert with a "hash expression" to ManageIQ] *****************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Creating alert Test Alert 01 failed: Api::BadRequestError: Resource expression needs be specified for creating a new alert_definitions"}
```

When trying to make this ansible call:
```
    - name: Add an alert with a "hash expression" to ManageIQ
      manageiq_alerts:
        state: present
        description: Test Alert 01
        options:
          notifications:
            email:
              to: ["example@example.com"]
              from: "example@example.com"
        resource_type: ContainerNode
        expression:
          eval_method: hostd_log_threshold
          mode: internal
          options: {}
        enabled: true
        manageiq_connection: "{{ miq_connection }}"

```

I did verify that deleting a non-existing alert now runs fine.

**Comment by elad661** [comment]
@joelddiaz You need https://github.com/ManageIQ/manageiq-api/pull/76 for this module to work, it's not merged yet.

**Comment by joelddiaz** [comment]
@elad661 i get the same thing with the 'miq' expression example

```
    - name: Add an alert with a "miq expression" to ManageIQ
      manageiq_alerts:
        state: present
        description: Test Alert 02
        options:
          notifications:
            email:
              to: ["example@example.com"]
              from: "example@example.com"
        resource_type: Vm
        expression_type: miq
        expression:
          and:
          - CONTAINS:
              tag: Vm.managed-environment
              value: prod
          - not:
            CONTAINS:
              tag: Vm.host.managed-environment
              value: prod
        enabled: true
        manageiq_connection: "{{ miq_connection }}"
```
results in
```
TASK [Add an alert with a "miq expression" to ManageIQ] ************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Creating alert Test Alert 02 failed: Api::BadRequestError: Resource expression needs be specified for creating a new alert_definitions"}
```

**Comment by joelddiaz** [comment]
@elad661 if i'm being perfectly honest, what i'd really like to create is this kind of alert (since this is what i repeatedly have to do every time i re-deploy MIQ) https://gist.github.com/joelddiaz/7b31dbca0552302047c8b5b861f7afdc

**Comment by elad661** [comment]
@joelddiaz at the moment you have to have the pull request merged in your local install to create any kind of alert (even ones using miq_expression) using the module. I can change it to be backwards-compatible for installations without that PR merged if needed - the only reason I didn't is because it makes the code a tiny bit less readable.

**Comment by joelddiaz** [comment]
@elad661 which PR are you referring to? this one (32136)? what i've tried to do up until now is check out your repo/branch and:
```
source ~/tmp/ansible-elad-manageiq-alerts/hacking/envsetup
ansible-playbook ./miq-alerts.yml
```

**Comment by elad661** [comment]
@joelddiaz the ansible module in this PR depends on a change to manageiq-api, PR  ManageIQ/manageiq-api#76 to work. You need to have that in your manageiq installation if you want this module to work.

**Comment by yaacov** [comment]
@elad661 notes about module requirements / compatibility -

A. If this module ( or some of it's functionality ) depend on another pull request:
    It should be noted on this PR description above.

B. If this module ( or some of it's functionality ) require minimal version of CFME:
    It also should be noted on this PR description above.

File: lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py]: this does belong in module_utils/manageiq.py as yaacov did.
perhaps send a small PR just adding this, it can get merged quickly, then both PRs can rely on it?

File: lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py:158]: I'm trying to think of a different class name, especially less confusable with `ManageIQAlert`.  But I don't have any good suggestion.

(this does fit the class naming in existing manageiq modules, matching the module name, but that's not such a good scheme IMHO, no reason to stick to it...)

File: lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py]: `url` vs `api_url` arbitrary.  perhaps call this `self.alerts_url`?

File: lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py:192]: what happens with older MIQ?  can you return to user a friendly message explaining it's not supported in this version?

File: lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py]: consider having miq_expression, hash_expression params.
- user would have to write just 1 param instead of currently needed 2 (type + expression)
- can natively document & validate each option differently
- this function mostly goes away :)  but this is least consideration.

File: lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py
Comment (cben) [inline lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py:149]: So this class can be initialized with both format we send and format we receive?
Neat! :heart:  Worth mentioning in docstring IMHO.

**Review by cben — COMMENTED** [review]
Looks good!
Some suggestions and nitpickss, none of them critical, feel free to disregard/push back.

File: lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py
Comment (elad661) [reply lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py]: I tried that in one of the early iterations, and it made the code more complicated because it'd need more conditions and extra validation, for example - I don't see a way to use Ansible's require_if to require one of two options but never both - so more conditions would be needed.

File: lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py
Comment (elad661) [reply lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py]: PR #32588 submitted

Comment (elad661) [reply lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py]: also, oops, we don't actually use this function here at all anymore, it was a leftover from an early iteration.

File: lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py
Comment (elad661) [reply lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py:158]: I can't think of a better name. We could add an extra word like "Handler" or something generic like that, but that's a big ugly.

File: lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py
Comment (elad661) [reply lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py:192]: See https://github.com/ansible/ansible/pull/32136#issuecomment-341520122 for an example of what happens with older MIQ.

The best thing we can do is catch that exception and change the error message to explain hash expression is not supported.

Comment (elad661) [reply lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py:192]: I've added a user-friendly error message for when the user tries to create/update a hash expression on a MIQ installation that does not support that.

File: lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py
Comment (elad661) [reply lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py:149]: Done

**Comment by elad661** [comment]
@cben I've addressed the review feedback. Let me know if anything else is needed.

**Review by cben — APPROVED** [review]
LGTM :+1:

don't think we should merge before API has landed.  (BTW, is API destined for upcoming gaprindashili release of ManageIQ?  or only master?)

**Comment by elad661** [comment]
> is API destined for upcoming gaprindashili release of ManageIQ? or only master?

I don't know, that's a question for @zakiva or @moolitayer

**Comment by yaacov** [comment]
> don't think we should merge before API has landed. (BTW, is API destined for upcoming gaprindashili release of ManageIQ? or only master?)

@elad661 I agree with @cben

@elad661 please address: 
https://github.com/ansible/ansible/pull/32136#issuecomment-341676047

**Comment by elad661** [comment]
@yaacov I've edited the description. Please note that I've made the module more compatible - https://github.com/ansible/ansible/pull/32136/commits/1c32abf67f6ec931ba196c48a6624f582d632293#diff-41ccedec6f2f48600442cd0ee9040380R206 it'll work with older versions of ManageIQ, but only for miq_expression style alerts. The API change is required only for hash_expression support.

**Comment by bazulay** [comment]
@gtanzillo PTAL ,
please note that this patch depends on an API patch (see https://github.com/ManageIQ/manageiq-api/pull/76)

**Comment by jctanner** [comment]
@yaacov @cben @bazulay is this module going to work if https://github.com/ManageIQ/manageiq-api/pull/76 is not merged?

**Comment by elad661** [comment]
@jctanner it'll work only for miq_expression type alerts, and trying to define a hash_expression one will cause an error.

**Comment by elad661** [comment]
Now that  ManageIQ/manageiq-api#76 was merged, this module is ready to be merged - just waiting for someone from ansible to review it.

**Comment by cben** [comment]
shipit

**Comment by yaacov** [comment]
shipit  🚢 💻

**Comment by elad661** [comment]
bot_status

**Comment by elad661** [comment]
Updated to remove state=list, as @resmo said this kind of thing is no longer acceptable in new modules.

@cben @yaacov please take another look.

**Comment by yaacov** [comment]
shipit  16x🐑

**Comment by resmo** [comment]
note the bot does not accept shipits with comments on the same line (anymore)...

shipit

**Comment by yaacov** [comment]
bot_status

**Comment by cben** [comment]
shipit

**Comment by elad661** [comment]
ping @resmo is there anything else needed to get this merged?

**Pull request merged**
